### PR TITLE
Ncnews 11

### PR DESCRIPTION
**Pull request by jgsn93**
add extra functionality to get articles to allow topic, sort by and order queries

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -32,15 +32,16 @@ describe("GET /api/topics", () => {
...
 
 describe("GET /api/articles", () => {
-  test("200: returns an object with an array with correctly formatted objects sorted by article ID in ascending order", () => {
+  test("200: returns an object with an array with correctly formatted objects sorted by most recent article in descending order by default", () => {
     return request(app)
       .get("/api/articles")
       .expect(200)
       .then(({ body }) => {
         const { articles } = body;
+        console.log(articles);
```
Comment (corkesw) [inline __tests__/app.test.js]: you should remove console logs

File: __tests__/app.test.js [hunk 2]
```diff
@@ -80,6 +81,42 @@ describe("GET /api/articles", () => {
         });
       });
   });
+
+  test("200: returns an object with an array with correctly formatted objects sorted any column in the table and either ascending or descending", () => {
+    return request(app)
+      .get("/api/articles?topic=mitch&sort_by=title&order=asc")
+      .expect(200)
+      .then(({ body }) => {
+        console.log(body);
```
Comment (corkesw) [inline __tests__/app.test.js]: you should remove console logs

File: __tests__/app.test.js [hunk 3]
```diff
@@ -80,6 +81,42 @@ describe("GET /api/articles", () => {
...
+        const { articles } = body;
+        expect(articles[0].title).toBe("A");
```
Comment (corkesw) [inline __tests__/app.test.js]: This assertion is not robust enough - I'd suggest using jest-sorted to check your assertion more throughly

File: __tests__/app.test.js [hunk 4]
```diff
@@ -80,6 +81,42 @@ describe("GET /api/articles", () => {
...
+
+  test("400: returns an error if sort by or order queries are not valid", () => {
+    return request(app)
+      .get("/api/articles?sort_by=durian")
+      .expect(400)
+      .then(({ body }) => {
+        expect(body.message).toBe("Invalid query");
+      });
+  });
+
```
Comment (corkesw) [inline __tests__/app.test.js]: You should separate out the testing for sort_by and order they represent separate logic.

Test wise we need:

 - Status 200, accepts `sort_by` query, e.g. `?sort_by=votes`
 - Status 200, accepts `order` query, e.g. `?order=desc`
 - Status 400. invalid `sort_by` query, e.g. `?sort_by=bananas`
 - Status 400. invalid `order` query, e.g. `?order=bananas`

File: models/articles.model.js [hunk 5]
```diff
@@ -1,39 +1,66 @@
...
+  //Query strings
+  let mainQuery = `SELECT articles.author, title, articles.article_id, topic, articles.created_at, articles.votes, CAST (COUNT(comment_id) AS INT) AS comment_count 
+  FROM articles 
+  JOIN comments ON articles.article_id = comments.article_id`;
+
+  const topicQuery = ` WHERE topic = $1`;
+  const groupByQuery = ` GROUP BY articles.article_id`;
+  const orderSortByQuery = ` ORDER BY ${sort_by} ${order}`;
+
+  const validQueries = [
```
Comment (corkesw) [inline models/articles.model.js]: you don't need to specify the table here (eg articles.author) and this is not the way that the query will be passed in by the user

File: __tests__/app.test.js [hunk 6]
```diff
@@ -80,6 +80,45 @@ describe("GET /api/articles", () => {
         });
       });
   });
+  test("200: returns an object with an array with correctly formatted objects sorted by any column in the table", () => {
```
Comment (jimstev2447) [inline __tests__/app.test.js:83]: you advent really tested here the order queries are working as intended, although you have handled the error cases

File: models/articles.model.js [hunk 7]
```diff
@@ -1,41 +1,61 @@
...
+    const query1 = db.query(
+      `${mainQuery} ${topicQuery} ${groupByQuery} ${orderSortByQuery};`,
+      [topic]
+    );
+    const query2 = db.query(`SELECT * FROM articles WHERE topic = $1;`, [
+      topic,
+    ]);
+
+    return Promise.all([query1, query2]).then((response) => {
+      if (response[0].rows.length !== 0 || response[1].rows.length !== 0) {
```
Comment (jimstev2447) [inline models/articles.model.js]: you could do some nicer variable naming maybe some restructuring here, response[0].rows isn't very clear what is going on

**Pull request merged**